PR: https://github.com/odoo/odoo/pull/214615

From: 1ee02f8a47d42d3ba3fd11ffcf8d9768ea17678e
From: Stephane Vanmeerhaeghe (stva)
Date: 2025-07-03 19:19:28

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'account.fiscal.position'

Total Changes: 177

[FIX] point_of_sale, pos_*: Fix loading behavior between self and POS

*: l10n_ar_pos, l10n_be_pos_sale, l10n_es_edi_tbai_pos, l10n_in_pos,
l10n_pe_pos, pos_discount, pos_event, pos_hr, pos_loyalty,
pos_online_payment_self_order, pos_restaurant, pos_sale, pos_self_order

Before this commit, the loading data behavior between self-ordering and
PoS was merged but not consistent. This led to issues when we try to
add some data to load for both self-ordering and PoS. At the moment, we
need to create two separate method to handle both cases.

Now with this commit, we have consistent methods which will load data
for both self-ordering and PoS. This will allow us to have a more
consistent behavior and avoid issues when adding data to load.

An explicit example is about `_server_version` object which is added to
the `pos.session` model and used when printing the receipt. Since
the session isn't always available in self-ordering, sometimes that
object isn't available, leading to issues when printing the receipt.

closes odoo/odoo#214615

Related: odoo/enterprise#87883
Signed-off-by: David Monnom (moda) <moda@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_self_order/models/__init__.py
+++ b/addons/pos_self_order/models/__init__.py
@@ -12,7 +12,6 @@ from . import product_tag
 from . import res_config_settings
 from . import pos_session
 from . import pos_load_mixin
-from . import account_fiscal_position
 from . import pos_preset
 from . import res_partner
 from . import res_country

--- a/addons/pos_self_order/models/account_fiscal_position.py
+++ b/None
@@ -1,8 +0,0 @@
-from odoo import models
-
-
-class AccountFiscalPosition(models.Model):
-    _inherit = 'account.fiscal.position'
-
-    def _load_pos_self_data(self, data):
-        return self._load_pos_data(data)

--- a/addons/pos_self_order/models/mail_template.py
+++ b/addons/pos_self_order/models/mail_template.py
@@ -5,9 +5,9 @@ class MailTemplate(models.Model):
     _inherit = ['mail.template', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('id', 'in', [preset['mail_template_id'] for preset in data['pos.preset']])]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id']

--- a/addons/pos_self_order/models/pos_config.py
+++ b/addons/pos_self_order/models/pos_config.py
@@ -285,43 +285,46 @@ class PosConfig(models.Model):
             'decimal.precision', 'uom.uom', 'pos.printer', 'pos_self_order.custom_link', 'restaurant.floor', 'restaurant.table', 'account.cash.rounding',
             'res.country', 'res.country.state', 'mail.template']
 
-    def load_self_data(self):
-        # Init our first record, in case of self_order is pos_config
-        config_fields = self._load_pos_self_data_fields(self.id)
-        response = {
-            'pos.config': self.env['pos.config'].search_read([('id', '=', self.id)], config_fields, load=False),
-        }
-        response['pos.config'][0]['_self_ordering_image_home_ids'] = self.self_ordering_image_home_ids.ids
-        response['pos.config'][0]['_self_ordering_image_background_ids'] = self.self_ordering_image_background_ids.ids
-        response['pos.config'][0]['_pos_special_products_ids'] = self._get_special_products().ids
-        response["pos.config"][0]["_self_ordering_style"] = {
-            "primaryBgColor": self.env.company.email_secondary_color,
-            "primaryTextColor": self.env.company.email_primary_color,
+    @api.model
+    def _load_pos_self_data_domain(self, data, config):
+        return [('id', '=', config.id)]
+
+    @api.model
+    def _load_pos_self_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
+        record = read_records[0]
+        record['_self_ordering_image_home_ids'] = config.self_ordering_image_home_ids.ids
+        record['_self_ordering_image_background_ids'] = config.self_ordering_image_background_ids.ids
+        record['_pos_special_products_ids'] = config._get_special_products().ids
+        record['_self_ordering_style'] = {
+            'primaryBgColor': self.env.company.email_secondary_color,
+            'primaryTextColor': self.env.company.email_primary_color,
         }
-        self.env['pos.session']._load_pos_data_relations('pos.config', response)
+        record['_self_order_pos'] = True
+        return read_records
+
+    def load_self_data(self):
+        response = {}
+        response['pos.config'] = self.env['pos.config']._load_pos_self_data_search_read(response, self)
 
-        # Classic data loading
         for model in self._load_self_data_models():
             try:
-                response[model] = self.env[model]._post_read_pos_self_data(self.env[model]._load_pos_self_data(response))
-                self.env['pos.session']._load_pos_data_relations(model, response)
+                response[model] = self.env[model]._load_pos_self_data_search_read(response, self)
             except AccessError:
-                response[model] = self.env[model]._load_pos_self_data_fields(self.id)
-
-                self.env['pos.session']._load_pos_data_relations(model, response)
+                response[model] = []
 
         return response
 
     def load_data_params(self):
         response = {}
-        fields = self._load_pos_self_data_fields(self.id)
+        fields = self._load_pos_self_data_fields(self)
         response['pos.config'] = {
             'fields': fields,
             'relations': self.env['pos.session']._load_pos_data_relations('pos.config', fields)
         }
 
         for model in self._load_self_data_models():
-            fields = self.env[model]._load_pos_self_data_fields(self.id)
+            fields = self.env[model]._load_pos_self_data_fields(self)
             response[model] = {
                 'fields': fields,
                 'relations': self.env['pos.session']._load_pos_data_relations(model, fields)

--- a/addons/pos_self_order/models/pos_load_mixin.py
+++ b/addons/pos_self_order/models/pos_load_mixin.py
@@ -6,17 +6,34 @@ class PosLoadMixin(models.AbstractModel):
     _inherit = "pos.load.mixin"
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        return self._load_pos_data_domain(data)
+    def _load_pos_self_data_search_read(self, data, config):
+        """ Search and return records to be loaded in the self """
+        if not config:
+            raise ValueError("config must be provided to search for PoS data.")
+
+        domain = self._load_pos_self_data_domain(data, config)
+        if domain is False:
+            return []
+
+        records = self.search(domain)
+        return self._load_pos_self_data_read(records, config)
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
-        return self._load_pos_data_fields(config_id)
+    def _load_pos_self_data_domain(self, data, config):
+        """ Return the domain used to filter records """
+        return self._load_pos_data_domain(data, config)
 
-    def _load_pos_self_data(self, data):
-        domain = self._load_pos_self_data_domain(data)
-        fields = self._load_pos_self_data_fields(data['pos.config'][0]['id'])
-        return self.search_read(domain, fields, load=False) if domain is not False else []
+    @api.model
+    def _load_pos_self_data_read(self, records, config):
+        """ Read specific fields from the given records """
+        if not config:
+            raise ValueError("config must be provided to read PoS data.")
 
-    def _post_read_pos_self_data(self, data):
-        return data
+        fields = self._load_pos_self_data_fields(config)
+        records = records.read(fields, load=False)
+        return records or []
+
+    @api.model
+    def _load_pos_self_data_fields(self, config):
+        """ Return the list of fields to be loaded """
+        return self._load_pos_data_fields(config)

--- a/addons/pos_self_order/models/pos_order.py
+++ b/addons/pos_self_order/models/pos_order.py
@@ -37,7 +37,7 @@ class PosOrder(models.Model):
     table_stand_number = fields.Char(string="Table Stand Number")
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
+    def _load_pos_self_data_domain(self, data, config):
         return [('id', '=', False)]
 
     @api.model

--- a/addons/pos_self_order/models/pos_payment_method.py
+++ b/addons/pos_self_order/models/pos_payment_method.py
@@ -9,8 +9,8 @@ class PosPaymentMethod(models.Model):
         pass
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        if data['pos.config'][0]['self_ordering_mode'] == 'kiosk':
+    def _load_pos_self_data_domain(self, data, config):
+        if config.self_ordering_mode == 'kiosk':
             return [('use_payment_terminal', 'in', ['adyen', 'stripe']), ('id', 'in', data['pos.config'][0]['payment_method_ids']) ]
         else:
             [('id', '=', False)]

--- a/addons/pos_self_order/models/pos_preset.py
+++ b/addons/pos_self_order/models/pos_preset.py
@@ -20,19 +20,18 @@ class PosPreset(models.Model):
 
     # will be overridden.
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        config_id = data['pos.config'][0]['available_preset_ids']
-        return ['|', ('id', '=', data['pos.config'][0]['default_preset_id']), '&', ('available_in_self', '=', True), ('id', 'in', config_id)]
+    def _load_pos_self_data_domain(self, data, config):
+        return ['|', ('id', '=', config.default_preset_id.id), '&', ('available_in_self', '=', True), ('id', 'in', config.available_preset_ids.ids)]
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
-        params = super()._load_pos_self_data_fields(config_id)
+    def _load_pos_self_data_fields(self, config):
+        params = super()._load_pos_self_data_fields(config)
         params.extend(['service_at', 'mail_template_id'])
         return params
     
     @api.model
-    def _load_pos_data_fields(self, config_id):
-        params = super()._load_pos_data_fields(config_id)
+    def _load_pos_data_fields(self, config):
+        params = super()._load_pos_data_fields(config)
         params.extend(['mail_template_id'])
         return params
 

--- a/addons/pos_self_order/models/pos_restaurant.py
+++ b/addons/pos_self_order/models/pos_restaurant.py
@@ -28,11 +28,11 @@ class RestaurantTable(models.Model):
             table.identifier = self._get_identifier()
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
+    def _load_pos_self_data_fields(self, config):
         return ['table_number', 'identifier', 'floor_id']
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
+    def _load_pos_self_data_domain(self, data, config):
         return [('floor_id', 'in', [floor['id'] for floor in data['restaurant.floor']])]
 
 
@@ -40,9 +40,9 @@ class RestaurantFloor(models.Model):
     _inherit = "restaurant.floor"
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
+    def _load_pos_self_data_fields(self, config):
         return ['name', 'table_ids']
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        return [('id', 'in', data['pos.config'][0]['floor_ids'])]
+    def _load_pos_self_data_domain(self, data, config):
+        return [('id', 'in', config.floor_ids.ids)]

--- a/addons/pos_self_order/models/pos_self_order_custom_link.py
+++ b/addons/pos_self_order/models/pos_self_order_custom_link.py
@@ -39,11 +39,11 @@ class Pos_Self_OrderCustom_Link(models.Model):
     sequence = fields.Integer("Sequence", default=1)
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        return [('pos_config_ids', 'in', data['pos.config'][0]['id'])]
+    def _load_pos_self_data_domain(self, data, config):
+        return [('pos_config_ids', 'in', config.id)]
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
+    def _load_pos_self_data_fields(self, config):
         return ['name', 'url', 'style', 'link_html', 'sequence']
 
     @api.depends("name", "style")

--- a/addons/pos_self_order/models/pos_session.py
+++ b/addons/pos_self_order/models/pos_session.py
@@ -14,11 +14,13 @@ class PosSession(models.Model):
         return data
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        return [('config_id', '=', data['pos.config'][0]['id']), ('state', '=', 'opened')]
+    def _load_pos_self_data_domain(self, data, config):
+        return [('config_id', '=', config.id), ('state', '=', 'opened')]
 
-    def _post_read_pos_data(self, data):
-        data[0]['_self_ordering'] = (
+    def _load_pos_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
+        record = read_records[0]
+        record['_self_ordering'] = (
             self.env["pos.config"]
             .sudo()
             .search_count(
@@ -31,10 +33,4 @@ class PosSession(models.Model):
             )
             > 0
         )
-        return super()._post_read_pos_data(data)
-
-    def _post_read_pos_self_data(self, data):
-        if data:
-            data[0]['_base_url'] = self.get_base_url()
-            data[0]['_self_order_pos'] = True
-        return super()._post_read_pos_self_data(data)
+        return read_records

--- a/addons/pos_self_order/models/product_product.py
+++ b/addons/pos_self_order/models/product_product.py
@@ -19,10 +19,9 @@ class ProductTemplate(models.Model):
         store=True
     )
 
-    def _load_pos_self_data(self, data):
-        domain = self._load_pos_self_data_domain(data)
-        fields = set(self._load_pos_self_data_fields(data['pos.config'][0]['id']))
-        config = self.env['pos.config'].browse(data['pos.config'][0]['id'])
+    def _load_pos_self_data_read(self, data, config):
+        domain = self._load_pos_self_data_domain(data, config)
+        fields = set(self._load_pos_self_data_fields(config))
         products = self.search_read(
             domain,
             fields,
@@ -44,10 +43,6 @@ class ProductTemplate(models.Model):
 
         return products
 
-    def _post_read_pos_self_data(self, data):
-        self._process_pos_self_ui_products(data)
-        return super()._post_read_pos_self_data(data)
-
     def _process_pos_self_ui_products(self, products):
         for product in products:
             product['_archived_combinations'] = []
@@ -56,14 +51,14 @@ class ProductTemplate(models.Model):
             product['image_128'] = bool(product['image_128'])
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
-        params = super()._load_pos_data_fields(config_id)
+    def _load_pos_data_fields(self, config):
+        params = super()._load_pos_data_fields(config)
         params += ['self_order_available']
         return params
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
-        domain = super()._load_pos_self_data_domain(data)
+    def _load_pos_self_data_domain(self, data, config):
+        domain = super()._load_pos_self_data_domain(data, config)
         return Domain.AND([domain, [('self_order_available', '=', True)]])
 
     @api.onchange('available_in_pos')
@@ -126,7 +121,7 @@ class ProductProduct(models.Model):
         for config in config_self:
             if config.current_session_id and config.access_token:
                 config._notify('PRODUCT_CHANGED', {
-                    'product.product': self.read(self._load_pos_self_data_fields(config.id), load=False)
+                    'product.product': self.read(self._load_pos_self_data_fields(config), load=False)
                 })
 
     def _can_return_content(self, field_name=None, access_token=None):

--- a/addons/pos_self_order/models/product_tag.py
+++ b/addons/pos_self_order/models/product_tag.py
@@ -11,5 +11,5 @@ class ProductTag(models.Model):
         return super()._can_return_content(field_name, access_token)
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
+    def _load_pos_self_data_domain(self, data, config):
         return [('visible_to_customers', '=', True)]

--- a/addons/pos_self_order/models/res_country.py
+++ b/addons/pos_self_order/models/res_country.py
@@ -6,6 +6,6 @@ class ResCountry(models.Model):
     _inherit = ['res.country', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_self_data_fields(self, config_id):
-        fields = super()._load_pos_self_data_fields(config_id)
+    def _load_pos_self_data_fields(self, config):
+        fields = super()._load_pos_self_data_fields(config)
         return fields + ["state_ids"]

--- a/addons/pos_self_order/models/res_partner.py
+++ b/addons/pos_self_order/models/res_partner.py
@@ -6,5 +6,5 @@ class ResPartner(models.Model):
     _inherit = 'res.partner'
 
     @api.model
-    def _load_pos_self_data_domain(self, data):
+    def _load_pos_self_data_domain(self, data, config):
         return False
